### DISCUSSION TITLE
Run in Google Colab

### DISCUSSION THREAD

**Discussion by monorural** [vladmandic/sdnext#81] [Ideas]
Hi there, I just came across this repo and am very excited to try it out. Is there a Google Colab available for running it?
If not, it would be awesome to have a notebook up to date in Google Colab, so that it can be run in the cloud.

This is an awesome example, very useful for me and lots of people who don't own strong enough GPUs to run it local: 

https://github.com/camenduru/stable-diffusion-webui-colab

Thanks, this looks very promising! 🙌

**Comment by vladmandic** [comment]
thanks, but i haven't made a collab notebook or tested if its compatible - just not enough time.

**Comment by econundrum** [comment]
If I can find sometime this weekend, I might have a try and see if I can get it working.

I'm no expert on colab either but I've used it a little bit, I have a card good enough to produce images but not for training so having dreambooth  built in along with other training methods is an advantage.

**Reply by monorural** [reply]
Thank you very much, I'm very interested to trying it out. Colab is an awesome tool, I've been using it since Disco Diffusion came out last year, and the amount of images I could create were huge thanks to it. Very reliable and having alternate Google accounts does the trick if you run out of computing power given for free.

**Comment by econundrum** [comment]
Colab only seems to support Python 3.9 at the moment so I'm dead in the water, though there may be hacks to get around this.

**Reply by vladmandic** [reply]
i'll relax requirements to allow python 3.9, but i cannot test it. check for update on github in ~1h.

**Reply by econundrum** [reply]
Sounds like they are upgrading colab to 3,10.x in the next couple of weeks anyway so might not be worth it.

**Comment by econundrum** [comment]
Got it running on colab, just want to tidy up a bit.

**Reply by monorural** [reply]
That's awesome!! I'll check it out today, thank you so much!

**Reply by monorural** [reply]
Can't find the link to the Colab, is it up? Let me know when it's ready so I can try it, thanks!

**Reply by LukeOvermind** [reply]
Hi, I am also highly interested!

**Reply by pimmen** [reply]
That's aweome, stranger! Could you be so kind and publish the notebook to your page? Really, really, _really_ awesome work!

**Reply by darthsnarkey** [reply]
I as well would LOVE to give this a run on Colab!!! I have it on my PC but I love the freedom colab gives.

**Reply by AI-Casanova** [reply]
I hacked together a colab for myself, but can't seem to get Codeformer working for some reason.

**Reply by AI-Casanova** [reply]
OMG fixing codeformer was a slog. 

Had to hack basicsr https://github.com/sczhou/CodeFormer/issues/193#issuecomment-1503410609
and update https://github.com/vladmandic/automatic/blob/c205251c65a87e8c55220e245f40b4a038e100ed/modules/codeformer_model.py#L77 to `                self.face_helper.face_detector.to(device)`

**Reply by vladmandic** [reply]
@AI-Casanova i see that issue is still open on codeformer repo? once its resolved, i can update this repo to use fixed version.
but patching it here is not possible since its an external repo that is cloned during setup.

**Reply by AI-Casanova** [reply]
This was more an informational note for anyone else running into the issue. 

I am wondering though if a try-except on your end might be appropriate to handle the change from `face_det` to `face_detector` here. https://github.com/sczhou/CodeFormer/commit/50489a6736ef6ac75f00582a646782049898fe43#diff-b834cf945a76e783c9d65f0aa7ef6e90c3b7c14c36e7bfb902bc6acbebd9289fR118

**Reply by vladmandic** [reply]
@AI-Casanova i'm open to it, can you try? i believe it would be in `modules/codeformer_model.py:send_model_to()`

**Reply by AI-Casanova** [reply]
`self.face_helper.face_detector.to(device)` absolutely works with the newest version of CodeFormers (I successfully post processed images with the latest commit after changing it), the problem comes if people use their older copy from A1111, so because of the breaking change we're tasked with backwards compatibility.

**Reply by vladmandic** [reply]
AFAIK, A1111 pinned CodeFormers version because it modified some core files in sd itself? if that is no longer the case, I can attempt to update CodeFormers packaged here, but i'd have to put it on my backlog.

in the meantime, can you try your idea with `try/except`?

**Reply by cenahum** [reply]
> Got it running on colab, just want to tidy up a bit.

Hi, did you get is running?  Could you give the colab?

**Reply by cenahum** [reply]
> AFAIK, A1111 pinned CodeFormers version because it modified some core files in sd itself? if that is no longer the case, I can attempt to update CodeFormers packaged here, but i'd have to put it on my backlog.
> 
> in the meantime, can you try your idea with `try/except`?

Hi Mandic, I would like to ask you, if you lived in Brazil in the 80's and part of 90's? There was a Mandic at that time in Brazil that lived in Sao Paulo and he had a BBS.

**Reply by monorural** [reply]
Edit: Bad news for * free * Colab users: Google doesn't allow running WebUI anymore on free accounts, they want only interactive code that is run directly from Google Colab. Automatic1111 is no longer an option for Google Colab :'(

**Reply by monorural** [reply]
Automatic1111 and any web interface of stable diffusion

**Reply by vladmandic** [reply]
> > AFAIK, A1111 pinned CodeFormers version because it modified some core files in sd itself? if that is no longer the case, I can attempt to update CodeFormers packaged here, but i'd have to put it on my backlog.
> > 
> > in the meantime, can you try your idea with `try/except`?
> 
> Hi Mandic, I would like to ask you, if you lived in Brazil in the 80's and part of 90's? There was a Mandic at that time in Brazil that lived in Sao Paulo and he had a BBS.

Nope, but I actually know who you're talking about vaguely as I used to run BBS as well loooong time ago and someone found that there are two of us 😂

**Reply by southbayjay** [reply]
> Bad news for Colab users: Google doesn't allow running WebUI anymore, they want only interactive code that is run directly from Google Colab. Automatic1111 is no longer an option for Google Colab :'(

Are you referring to the free version?? I'm pretty sure I was just using it yesterday on A1111 but on the paid version. Well I'm going to give it a try, maybe something changed.

**Reply by AI-Casanova** [reply]
There is currently a string search now in colab for "Stable-diffusion-webui" so this fork *should* actually be safe for now.

**Reply by southbayjay** [reply]
> Bad news for Colab users: Google doesn't allow running WebUI anymore, they want only interactive code that is run directly from Google Colab. Automatic1111 is no longer an option for Google Colab :'(

I'm running A1111 right now on Colab, not sure what you are referring to but it is functional. 🤷🏻‍♂️

**Reply by monorural** [reply]
*Edit* only affects free users.

I'm referring to this: 

https://www.reddit.com/r/StableDiffusion/comments/12tj22r/official_statement_from_the_person_in_charge_of/
https://www.reddit.com/r/StableDiffusion/comments/12trnw9/now_that_they_started_banning_stable_diffusion_on/

Google is starting to ban the usage of WebUIs with Google Colab, aiming mostly at stable-diffusion-webui because the usage just keeps growing and got huge

**Reply by AI-Casanova** [reply]
*free usage

**Reply by monorural** [reply]
you are right, my mistake

**Reply by cenahum** [reply]
They will not ban the users, but they can shut down your app in a minute or 2, maybe you can run for a while, this only happens because people do not respect the free thing, there is a lot of this kind of cases, they abuse and everybody pay, Midjourney is another example, it's a shame, but is reality. Anyway, people will now move to Kaggle and Gradiente and soon this will do the same. Anyway, there are some cheap options that we can use, or we can install in a desktop or in a notebook.

**Reply by cenahum** [reply]
Ok, I see that they change the mind, and they can Ban user if try to use SD webui, but I read that they will let use SD with Gradio.

**Comment by southbayjay** [comment]
Hey folks! Is there any update on the Colab notebook for this?

**Reply by spreck** [reply]
I found this code on some reddit comment the other day and it seemed to work fine:

#@title Connect to Google Drive
from google.colab import drive
drive.mount('/content/drive')

#@title V2 - Clone Vladmandic SD
%cd /content/
!git clone https://github.com/vladmandic/automatic

#@title Extensions install (optional)
%cd /content/automatic/extensions/
!git clone https://github.com/fkunn1326/openpose-editor
!git clone https://github.com/mcmonkeyprojects/sd-dynamic-thresholding
!git clone https://github.com/kohya-ss/sd-webui-additional-networks.git
!git clone https://github.com/Klace/stable-diffusion-webui-instruct-pix2pix
!git clone https://github.com/KohakuBlueleaf/a1111-sd-webui-locon
!git clone https://github.com/opparco/stable-diffusion-webui-composable-lora
!git clone https://github.com/opparco/stable-diffusion-webui-two-shot
!git clone https://github.com/nonnonstop/sd-webui-3d-open-pose-editor
!git clone https://github.com/richrobber2/canvas-zoom
!git clone https://github.com/hako-mikan/sd-webui-regional-prompter
!git clone https://github.com/ilian6806/stable-diffusion-webui-state
!git clone https://github.com/jtydhr88/sd-canvas-editor

#@title Controlnet 1.1 models install (optional, need huggingface token)
user_token = "" #@param {type:"string"}
user_header = f"\"Authorization: Bearer {user_token}\""
%cd /content/
!mkdir -p ControlNet
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11e_sd15_ip2p.pth -O /content/ControlNet/control_v11e_sd15_ip2p.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11e_sd15_ip2p.yaml -O /content/ControlNet/control_v11e_sd15_ip2p.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11e_sd15_shuffle.pth -O /content/ControlNet/control_v11e_sd15_shuffle.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11e_sd15_shuffle.yaml -O /content/ControlNet/control_v11e_sd15_shuffle.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11f1p_sd15_depth.pth -O /content/ControlNet/control_v11f1p_sd15_depth.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11f1p_sd15_depth.yaml -O /content/ControlNet/control_v11f1p_sd15_depth.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_canny.pth -O /content/ControlNet/control_v11p_sd15_canny.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_canny.yaml -O /content/ControlNet/control_v11p_sd15_canny.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_inpaint.pth -O /content/ControlNet/control_v11p_sd15_inpaint.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_inpaint.yaml -O /content/ControlNet/control_v11p_sd15_inpaint.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_lineart.pth -O /content/ControlNet/control_v11p_sd15_lineart.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_lineart.yaml -O /content/ControlNet/control_v11p_sd15_lineart.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_mlsd.pth -O /content/ControlNet/control_v11p_sd15_mlsd.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_mlsd.yaml -O /content/ControlNet/control_v11p_sd15_mlsd.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_normalbae.pth -O /content/ControlNet/control_v11p_sd15_normalbae.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_normalbae.yaml -O /content/ControlNet/control_v11p_sd15_normalbae.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_openpose.pth -O /content/ControlNet/control_v11p_sd15_openpose.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_openpose.yaml -O /content/ControlNet/control_v11p_sd15_openpose.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_scribble.pth -O /content/ControlNet/control_v11p_sd15_scribble.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_scribble.yaml -O /content/ControlNet/control_v11p_sd15_scribble.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_seg.pth -O /content/ControlNet/control_v11p_sd15_seg.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_seg.yaml -O /content/ControlNet/control_v11p_sd15_seg.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_softedge.pth -O /content/ControlNet/control_v11p_sd15_softedge.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15_softedge.yaml -O /content/ControlNet/control_v11p_sd15_softedge.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15s2_lineart_anime.pth -O /content/ControlNet/control_v11p_sd15s2_lineart_anime.pth
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11p_sd15s2_lineart_anime.yaml -O /content/ControlNet/control_v11p_sd15s2_lineart_anime.yaml
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11u_sd15_tile.pth -O /content/ControlNet/control_v11u_sd15_tile.pt
!wget --no-clobber --header={user_header} https://huggingface.co/lllyasviel/ControlNet-v1-1/resolve/main/control_v11u_sd15_tile.yaml -O /content/ControlNet/control_v11u_sd15_tile.yaml

#@title Launch Vladmandic webui (GPU mode)
%cd /content/automatic/
!COMMANDLINE_ARGS="--no-hashing --share --noupdate --skip-extensions --gradio-auth login:pass --ckpt /content/drive/MyDrive/SDmodels/clarity_19-pruned.safetensors --ui-config-file /content/drive/MyDrive/1my/vlad-ui-config-my.json --ui-settings-file /content/drive/MyDrive/1my/vlad-config-my.json" REQS_FILE="requirements.txt" python launch.py

**Reply by mykeehu** [reply]
Yes, it's based on [my colab](https://github.com/vladmandic/automatic/issues/192#issuecomment-1516692776), since then I removed the `--noupdate --skip-extensions` options, because need the extension update.

But what I noticed is problematic on Colab:
- I can't change the model, it drops the Gradio connection
- I can't load SD 1.5 inpainting model, because as soon as I upload an image, the colab stops working.

But the image generation works, and I can change the model by changing the path in --ckpt.

**Comment by RimorCosmicam** [comment]
Hey there!

I've been using it for a while with Colab but didn't see the request, I have a single-cell colab to run/update it, however, you'll have to self-host the models on GDrive (I sync it with my computer, so it's simpler for me to add models and Loras), just run it, add your models to their respective folders, and you're good to go.

[Vladmandic A1111](https://colab.research.google.com/drive/1soFm1YPjidro_FqC9opdgSZSJIFJ0HnK?usp=sharing)

**Reply by vladmandic** [reply]
@RimorCosmicam thanks for sharing!

**Reply by wordbrew** [reply]
beautiful. Just saved me a ton of time. :)

**Reply by southbayjay** [reply]
Fantastic!! I'm jumping on Colab now to give this a try! Thank you!

**Reply by abragano** [reply]
what is a normal time for this to run? When I run it, takes 12 - 15 minutes.

**Reply by ash-bluepollution95** [reply]
Normal time for those that pre-install to your drive DO take loinger than the ones that are complete and only in the notebook, don't fear the ones that are on your drive :P I use a mixture of notebooks, both NOcrypt, Lastben (which saves EVERYTHING to my drive and it's convinent XD) - May test this one soon!

**Reply by cenahum** [reply]
Hi @RimorCosmicam
Thank you very much for sharing

**Reply by BinaryQuantumSoul** [reply]
> Hey there!
> 
> 
> 
> I've been using it for a while with Colab but didn't see the request, I have a single-cell colab to run/update it, however, you'll have to self-host the models on GDrive (I sync it with my computer, so it's simpler for me to add models and Loras), just run it, add your models to their respective folders, and you're good to go.
> 
> 
> 
> [Vladmandic A1111](https://colab.research.google.com/drive/1soFm1YPjidro_FqC9opdgSZSJIFJ0HnK?usp=sharing)

Sadly it doesn't work. I have this error:
```

ImportError: cannot import name 'get_device' from 'basicsr.utils.misc'
(/usr/local/lib/python3.10/dist-packages/basicsr/utils/misc.py)

````

**Reply by Protonk** [reply]
This will run on the very first update, but subsequent runs get stuck after listing the Python version. I have no idea why this happens, no error messages are emitted, it just stalls before the `02:58:53-632197 INFO     Version: 22e58809 Sat May 6 08:51:23 2023 -0400` string.

**Reply by AI-Casanova** [reply]
@Protonk just figured out that that is an issue with git since they upgraded colab to python 3.10

Add --skip-git and you'll be golden.

**Reply by Protonk** [reply]
[AI-Casanova](https://github.com/AI-Casanova) This works to solve that problem, though after a few server restarts the process fails as it cannot find clip.

**Comment by wa2el** [comment]
first of all thanks very much as everything is almost working perfectly

my problem is that CodeFormers isn't working
```
16:56:00-752061 INFO     Installing repositories                                
16:56:47-138300 ERROR    Error running git: None / -C                           
                         "/content/drive/MyDrive/A1113/repositories/CodeFormer" 
                         checkout c5b4593074ba6214284d6acd5f1719b6c5d739af      
16:56:47-140723 ERROR    Local changes detected: check setup.log for details
```
any idea how to solve this?

**Reply by cenahum** [reply]
Hi, I'm also facing some errors:   Error running git: None / -C                           
                         "/content/drive/MyDrive/A1111/repositories/CodeFormer" 
                         checkout 
ERROR    Local changes detected: check setup.log for details 

ImportError: cannot import name 'get_device' from 'basicsr.utils.misc' 
(/usr/local/lib/python3.10/dist-packages/basicsr/utils/misc.py)

**Reply by BinaryQuantumSoul** [reply]
Same issue here

**Reply by AI-Casanova** [reply]
CodeFormer has a couple annoying issues, but they are solvable. 

`/usr/local/lib/python3.10/dist-packages/basicsr/utils/misc.py` needs to be modified, as the CodeFormer people modified the package away from the one you can pip install.

**Reply by AI-Casanova** [reply]
These are the two fixes needed

https://github.com/vladmandic/automatic/discussions/81#discussioncomment-5691801

**Comment by vladmandic** [comment]
you may want to check out <https://colab.research.google.com/drive/126cDNwHfifCyUpCCQF9IHpEdiXRfHrLN>
and report issues directly at <https://github.com/Linaqruf/sd-notebook-collection> - in general, those are best google colab notebooks and he offered to add one for this repo!

**Reply by BinaryQuantumSoul** [reply]
Nice, keep up with collabs :)

**Reply by BinaryQuantumSoul** [reply]
> you may want to check out https://colab.research.google.com/drive/126cDNwHfifCyUpCCQF9IHpEdiXRfHrLN and report issues directly at https://github.com/Linaqruf/sd-notebook-collection - in general, those are best google colab notebooks and he offered to add one for this repo!

Shouldn't you also ask on https://github.com/TheLastBen/fast-stable-diffusion ?
He's waiting for the repo to "be well established" before adding it

**Reply by vladmandic** [reply]
author says its on the way: <https://github.com/TheLastBen/fast-stable-diffusion/issues/2062#issuecomment-1528133889>

**Reply by wordbrew** [reply]
Honestly, I'm not hopeful for TheLastBen properly incorporating vladmandic. I have used Fast A1111 on colab for a few months now and it actually boots and runs slower than vladmandic on colab. I can't imagine TheLastBen's customizations to A1111 will improve vladmandic more than anything you've already done.

I've got a decent colab notebook built out that follows much of the same build out that TheLastBen has for his colab. Just need to find the time to incorporate the Controlnet models and I'm good.

**Reply by eucdee** [reply]
> you may want to check out https://colab.research.google.com/drive/126cDNwHfifCyUpCCQF9IHpEdiXRfHrLN and report issues directly at https://github.com/Linaqruf/sd-notebook-collection - in general, those are best google colab notebooks and he offered to add one for this repo!

this colab is great out of the box, ive tried it yesterday with very simple promts and got stunning results 🗡️ 
so ive installed the github version localy and sadly its way different with many options still disabled.
sure i dont even know what 97% of these options really do yet but it would be nice to have them ;)
the authors notebook Cagliostro Colab UI is amazing aswell... i believe it uses anapnoes fork
i havent had time to check anapnoes fork but it would be nice to know the difference between the two.
ive tried setting up my local install to work like the colabs but i guess it would be better to try and run the notebook localy.
thats what will do now...

**Comment by wordbrew** [comment]
If anyone wants to use a colab, here is the one I built that Frankensteined some of TheLastBen's colab cells. It has some errors that I haven't bothered to fix yet (like the one @AI-Casanova solved for Codeformers). But it should work.

A few things to note:
1. This is for 1.5 SD. I took out anything pertaining to 2.x
2. I have not tested the ControlNet cell simply because I already have the 20gig files installed elsewhere. It should work though...

https://github.com/wordbrew/vladcolab/blob/main/Vladmandic_A1111.ipynb

I'm new(er) to Git, so sorry if this isn't right way to share the colab. I had built a gist, but it doesn't appear to allow you to easily copy the colab to your gdrive. This git repo does. I probably won't be maintaining it though.

**Comment by djwashout** [comment]
Humm. I'm new to all of this but let me see if i understand. From what i have gathered a notebook is just a notebook that runs commands in colab, (like on old bat file) and a repository is the actual code online in github that can install or execute? Sorry last time i had to learn tech was dos 5.0 running win.exe lol. its been a while! ...but if im right above. then damn. I just need to google the hell out of how to write a google colab notebook and im set. I was under the impression vladmandic had to be installed on a local machine...

**Reply by wordbrew** [reply]
Just view the thread. There are multiple colab notebooks already shared.

But yes, generally speaking, you're right.

And don't bother googling. Use ChatGPT to help you code the notebook if you want.

**Comment by southbayjay** [comment]
You can install it on a local machine, it was intended for that purpose
originally but some of us don’t have the compute power to run it locally so
we use other sources like Google Colab. If you have a capable pc to run it
then jump on YouTube and look up Olivio tutorials. He did a pretty
straightforward tutorial on installing it locally.
